### PR TITLE
Reapply compile in `Fabric.setup()` by default

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Switches the default for `Fabric.setup(..., reapply_compile=True|False)` to `True`. 
Follow up to #19280

I experimented with several (FSDP) models in lit-gpt to show the impact by modifying the `tinyllama.py` script with different models, recording the average iteration speed:

- TinyLlama 1.1B:  436 ms -> 436 ms, 35.53 -> 35.53 GB
- Pythia 160m: 231 ms -> 232 ms, 31.44 GB -> 31.44 GB
- Open Llama 3B (bf16-true, act. ckpt): 478 ms (+/- 1) -> 480 ms (+/- 2), 27.09 GB -> 27.07 GB
- Phi, StableLM, StableCode did not compile

(in each experiment, maximized batch size to push GPU to 99% util)
PyTorch 2.2, CUDA 12.1, 8xA100

Conclusion: I don't see a considerable impact with the few models I tested.

----
📚 Documentation preview 📚: https://pytorch-lightning--19382.org.readthedocs.build/en/19382/

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

File: docs/source-fabric/advanced/compile.rst [hunk 1]
```diff
@@ -275,11 +275,11 @@ In the future, this option will become the default.
     # Compile the model
     model = torch.compile(model)
 
-    # Default: `fabric.setup()` will not reapply the compilation over DDP/FSDP
-    model = fabric.setup(model, _reapply_compile=False)
+    # Default: `fabric.setup()` will configure compilation over DDP/FSDP for you
+    model = fabric.setup(model, reapply_compile=True)
```
Comment (carmocca) [inline docs/source-fabric/advanced/compile.rst:279]: Should we keep the argument protected? We don't want users to think they need to touch this. It should only be an emergency escape hatch

**Review by carmocca — COMMENTED** [review]
Flip it!

File: docs/source-fabric/advanced/compile.rst (hunk 1 above)
Comment (awaelchli) [reply docs/source-fabric/advanced/compile.rst:279]: That's also fine with me. Would you then still keep the docs section as it is right now?

Comment (carmocca) [reply docs/source-fabric/advanced/compile.rst:279]: Sure, maybe move it to the bottom

File: docs/source-fabric/advanced/compile.rst [hunk 2]
```diff
@@ -255,17 +256,33 @@ Naturally, the tradoff here is that it will consume a bit more memory.
...
+For these reasons, we recommend that you don't waste too much time trying to apply ``torch.compile`` during development, and rather evaluate its effectiveness toward the end when you are about to launch long-running, expensive experiments.
+Always compare the speed and memory usage of the compiled model against the original model!
+
+
 ----
 
 
-*******************************************************
-(Experimental) Apply torch.compile over FSDP, DDP, etc.
-*******************************************************
```
Comment (awaelchli) [inline docs/source-fabric/advanced/compile.rst:263]: switched the order of these two sections, no other changes

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Pull request merged**